src/inc/til/at.h

```diff
@@ -3,19 +3,82 @@
 
 #pragma once
 
+// The at function declares that you've already sufficiently checked that your array access
+// is in range before retrieving an item inside it at an offset.
+// This is to save double/triple/quadruple testing in circumstances where you are already
+// pivoting on the length of a set and now want to pull elements out of it by offset
+// without checking again.
+// gsl::at will do the check again. As will .at(). And using [] will have a warning in audit.
+
 namespace til
 {
-    // The at function declares that you've already sufficiently checked that your array access
-    // is in range before retrieving an item inside it at an offset.
-    // This is to save double/triple/quadruple testing in circumstances where you are already
-    // pivoting on the length of a set and now want to pull elements out of it by offset
-    // without checking again.
-    // gsl::at will do the check again. As will .at(). And using [] will have a warning in audit.
-    template<class T>
-    constexpr auto at(T& cont, const size_t i) -> decltype(cont[cont.size()])
+    // Routine Description:
+    // - Takes a reference to an array of constant data and returns the element at the given offset.
+    //   NOTE: The function relies on a sufficient check that the access is in range.
+    // Arguments:
+    // - array - Reference to constant data in an array.
+    // - index - Offset of the element to be returned.
+    // Return Value:
+    // - Element in the array at the offset given by the index parameter.
+    template<class T, class U, const size_t N>
+    constexpr auto at(const T (&array)[N], const U index) -> typename std::enable_if<std::is_integral<U>::value, decltype(array[0])>::type

```

Comment:
If we replace `T (&array)[N]` and `const T (&array)[N]` with `T (&&array)[N]`, can we get this down to one specialization? i don't mind too much if not :)

Reply:
I'll definitely try. Thanks!

Reply:
@DHowett-MSFT Unfortunately this still leads to an array to pointer decay. I'll remove the array overloads and work around using the pointer overload even if that was definitely not my intention 😥 Sorry for all the commits. But I can't reproduce that locally.